src/gui/DatabaseTabWidget.cpp

```diff
@@ -510,272 +369,140 @@ void DatabaseTabWidget::changeDatabaseSettings()
 {
+    if (count() == 0) {
+        return false;
+    }
+
     if (index == -1) {
         index = currentIndex();
     }
 
-    return indexDatabaseManagerStruct(index).readOnly;
+    auto db = databaseWidgetFromIndex(index)->database();
+    return db && db->isReadOnly();
 }
 
-bool DatabaseTabWidget::canSave(int index)
+bool DatabaseTabWidget::isModified(int index) const
 {
+    if (count() == 0) {
+        return false;
+    }
+
     if (index == -1) {
         index = currentIndex();
     }
 
-    const DatabaseManagerStruct& dbStruct = indexDatabaseManagerStruct(index);
-    return dbStruct.modified && !dbStruct.readOnly;
+    auto db = databaseWidgetFromIndex(index)->database();
+    return db && db->isModified();
 }
 
-bool DatabaseTabWidget::isModified(int index)
+bool DatabaseTabWidget::canSave(int index) const
 {
-    if (index == -1) {
-        index = currentIndex();
-    }
-
-    return indexDatabaseManagerStruct(index).modified;
+    return !isReadOnly(index) && isModified(index);
 }
 
-QString DatabaseTabWidget::databasePath(int index)
+bool DatabaseTabWidget::hasLockableDatabases() const
 {
-    if (index == -1) {
-        index = currentIndex();
+    for (int i = 0, c = count(); i < c; ++i) {
+        if (databaseWidgetFromIndex(i)->currentMode() != DatabaseWidget::Mode::LockedMode) {
+            return true;
+        }
     }
-
-    return indexDatabaseManagerStruct(index).fileInfo.absoluteFilePath();
+    return false;
 }
 
-void DatabaseTabWidget::updateTabName(Database* db)
+/**
+ * Get the tab's (original) display name without platform-specific
+ * mangling that may occur when reading back the actual widget's \link tabText()
+ *
+ * @param index tab index
+ * @return tab name
+ */
+QString DatabaseTabWidget::tabName(int index)
 {
-    int index = databaseIndex(db);
-    Q_ASSERT(index != -1);
+    if (index == -1 || index > count()) {
+        return "";
+    }
 
-    const DatabaseManagerStruct& dbStruct = m_dbList.value(db);
+    auto* dbWidget = databaseWidgetFromIndex(index);
+
+    auto db = dbWidget->database();
+    Q_ASSERT(db);
+    if (!db) {
+        return "";
+    }
 
     QString tabName;
-    QString fileName;
 
-    if (dbStruct.fileInfo.exists()) {
+    if (!db->filePath().isEmpty()) {
+        QFileInfo fileInfo(db->filePath());
+
         if (db->metadata()->name().isEmpty()) {
-            tabName = dbStruct.fileInfo.fileName();
+            tabName = fileInfo.fileName();
         } else {
             tabName = db->metadata()->name();
         }
 
-        fileName = dbStruct.fileInfo.fileName();
-        setTabToolTip(index, dbStruct.fileInfo.absoluteFilePath());
+        setTabToolTip(index, fileInfo.absoluteFilePath());
     } else {
         if (db->metadata()->name().isEmpty()) {
-            tabName = tr("New database");
+            tabName = tr("New Database");
         } else {
-            tabName = tr("%1 [New database]", "tab modifier").arg(db->metadata()->name());
```

Comment:
To be consistent with above, should be DatabaseWidget::Mode::LockedMode

Reply:
I think it's like that everywhere and I added some new ones with Mode.